### ISSUE TITLE
(roofit) RooHistPdf providing a wrong numerical value, root 6.18.00

### ISSUE THREAD

**Issue by marcescalier** [root-project/root#7991]
Hello,

*Two histograms of m_yy are filled with pseudo-data :
-one for signal, filled mainly in the region of 125 GeV (no long tails on purpose, in order to show clearly the problem) (total in the 55 bins=1.58019)
-one for the continuum, flat with a value of 40 (total in the 55 bins=2160)

*For each of these histograms, a RooDataHist are filled. They will serve to construct the binned pdf

*A RooDataHist is constructed for the total : this is the pseudo-data.

*The number or signal and bkg are counted and represent two variables (no fit, just counting what was filled)

*From the two RooDataHist, an "option" (flag) renormalizes or does not renormalize the RooDataHist.

*From these two RooDataHist, [two](url) HistPdf are filled.

*Then, three models of pdf are constructed :
a pdf total : n_s * pdf_signal + n_b * pdf_continuum

a pdf signal only : pdf_signal
a pdf continuum only : pdf_continuum

a pdf n_s * pdf_signal : n_s * pdf_signal
a pdf n_continuum * pdf_continuum

Then, the first bin of the pseudo-data is printed : 40 (indeed, there is no signal in this bin)

*Then, we compare this to the value of the pdf
-->The two variantes  of option (normalize or not renormalize the RooDataHist) give a wrong value for the pdf.

Indeed :
*In the case where we don't normalize the RooDataHist,
then the pdf total in this (example of) bin is wrong (by a factor 2160 / (2160+1.58019)

*In the case where we do not normalize the RooDataHist,
-then the pdf n_continuum * pdf_continuum is wrong (by a factor 2160)

-then the pdf pdf_continuum is wrong : by a factor (2160+1.58019) / 2160

I add a pdf illustrating the problem.

I put here a minimum program that reproduces the problem :
/afs/cern.ch/work/e/escalier/public/ForOthers/ForRootExperts/ProblemRooHistPdf%

Thank you

(Remark : the reason of the apparently strange list of steps is in order to do a much more complex study, which is not the purpose of the problem reported here.)

**Comment by marcescalier** [comment]
pdf for illustration [Problem_root.pdf](https://github.com/root-project/root/files/6368027/Problem_root.pdf)

**Comment by guitargeek** [comment]
Hi @marcescalier!

I have studied your code and slides (thanks for all the details as always!)

However, I can't see the bug here, i.e. which values are not like you would expect. You are right when you say "RooHistPdfisexpectedto manage the normalization, fromthe RooDataHist(sowhateverI normalizeor don’tnormalize, it shoulddo the job)".

First of all, I think there was a little problem in the script. When you fill the histogram, I think you intended to do:
```C++
for (int index_bin=1;index_bin<=55;index_bin++)
```
You had `<55` in the code, but with this, the last bin of the continuum would be empty.

Now, I'll go over the output of your script for both `option_normalize_hist=1` and `option_normalize_hist=0`, and I'll try to explain why everything makes sense to me. Let's start with `option_normalize_hist=1`:
```
----------------------------
prepare model for index_process=0, signal
----------------------------
prepare model for index_process=1, Continuum
nb_events=2201.5801859878
RooRealVar::n_signal = 1.58019  L(0 - 10000) 
RooRealVar::n_Continuum = 2200  L(0 - 10000) 
nb_entries=55
-----
data : m_yy=105.5, value data=40
value_pdf total=40
n_times_pdf_signal->getVal()*bin_width=0
n_times_pdf_continuum->getVal()*bin_width=0.0181818
pdf signal only : 0
pdf bkg only : 40.0287
```
1. `value_pdf total=40`: I think we agree here that 40 is correct, because there is no signal in the first bin
2. `n_times_pdf_signal->getVal()*bin_width=0`: trivial
3. `n_times_pdf_continuum->getVal()*bin_width=0.0181818` : here `n_times_pdf_continuum` is a `RooAddPdf`, which automatically normalizes itself. You create it from the continuum with 55 bins and bin width one, so I'd expect 1/55 which is indeed 0.0181818
4. `pdf signal only : 0` trivial
5. `pdf bkg only : 40.0287`: this is `nb_events` multiplied by your manually normalized background `RooDataHist`. `nb_events` is 2200 + 1.58019, and the weight of your manually normalized `RooDataHist` will again be 1/55. So Indeed I expect (2200 + 1.58019)/55 = 40.0287

Now the output with with `option_normalize_hist=0`:
```
----------------------------
prepare model for index_process=0, signal
----------------------------
prepare model for index_process=1, Continuum
nb_events=2201.5801859878
RooRealVar::n_signal = 1.58019  L(0 - 10000) 
RooRealVar::n_Continuum = 2200  L(0 - 10000) 
nb_entries=55
-----
data : m_yy=105.5, value data=40
value_pdf total=0.0181688
n_times_pdf_signal->getVal()*bin_width=0
n_times_pdf_continuum->getVal()*bin_width=0.0181818
pdf signal only : 0
pdf bkg only : 40
```
1. `value_pdf total=0.0181688`:  Before it was 40, now it's 40/`nb_events` = 40/(2200 + 1.58019) = 0.0181688.
2. `n_times_pdf_signal->getVal()*bin_width=0` trivial
3. `n_times_pdf_continuum->getVal()*bin_width=0.0181818` : As expected again 1/55, because the `RooAddPdf` always normalizes itself.
4. `pdf signal only : 0` : trivial
5. `pdf bkg only : 40`: you print the counts of your unnormalized background histogram, so 40 is correct here.

In conclusion, I can't see a bug in RooFit here as far as I can tell. Maybe you were caught by surprise that the pdfs normalize themselves?

**Comment by marcescalier** [comment]
Dear Jonas.

A huge thank you for your investigation on this, and kind explanation. Almost everything on what you explained is clear to me. But let me give you a tiny point where I don't completely understand.

**About the <55 instead of <=55. You are right : I made a mistake there in my minimum example. But the principle of the comment stay. I made the fix in order to "synchronize" with the numerical values that you gave in your kind answer (I deduced that you made the fix).

-->ok for that

**Case option_normalize_hist=1.

Your comment : "n_times_pdf_continuum->getVal()*bin_width=0.0181818 : here n_times_pdf_continuum is a RooAddPdf, which automatically normalizes itself. You create it from the continuum with 55 bins and bin width one, so I'd expect 1/55 which is indeed 0.0181818"
-->I disagree : my pdf is not just "pdf_continuum" : I put *n* times pdf_continuum where n is here the number of bkg : n=2200.
So I would expect that n times pdf should be 2200 * (1/55) = 40

**n_times_pdf_continuum is a RooAddPdf (with one element in the addition) with one parameter of yield, and one parameter of shape, so it is a extended pdf : n_bkg * times a pdf.**


**Case option_normalize_hist=0.
From your log, you changed a bit my code since in my code, I was putting, for the case not normalize :
    if (option_normalize_hist)
      value_pdf=nb_events*pdf->getVal()*bin_width;
    else
      value_pdf=pdf->getVal()*bin_width;

But I understood your point (and it is clear)

**Comment by guitargeek** [comment]
Hi! Yes that's true that your `n_times_pdf_continuum` is an extended pdf (just like in the [first RooFit example](https://root.cern/manual/roofit/)). But still, if you get the value of a pdf, it will be normalized. The coefficient information will only be used in the extended term when you do a fit or for plotting.

If you want to use an addition class that doesn't normalize itself, you can also take a look at [RooRealSumFunc](https://root.cern.ch/doc/v614/classRooRealSumFunc.html).

**Comment by marcescalier** [comment]
Thank you.

So if I understand correctly : 
for n_s times RooHistPdf, it behaves as for the case RooHistPdf for the method : getVal() 
-> OK : it gives a result "normalized to 1".

But if we decide *not to normalize* the RooDataHist used to create the RooHistPdf,
the RooHistPdf -> getVal() will return a value normalized to the number of events of the RooDataHist : so in this case it gives a *non-normalized result*

This is strange since it is stated : "The histogram distribution is explicitly normalized by RooHistPdf and can have an arbitrary number of real or discrete dimensions. "

Indeed, with my minimum example where you may put the option option_normalize_hist=0 (with the fix that you kindly found of <=55 bins)
we get :
value_pdf total=39.9713
which represents : pdf->getVal()*bin_width

so in this case, we don't need to multiply by the number of events to obtain a prediction !!
The value of the pdf is not inside [0, 1]

So, depending whether the RooDataHist that was used in order to construct the RooHistPdf, was normalized or not normalized, the RooHistPdf->getVal() returns a normalized or a not normalised result,
in contradiction with the sentence that roofit is expected to do it : the sentence : "The histogram distribution is explicitly normalized by RooHistPdf and can have an arbitrary number of real or discrete dimensions. "

**Comment by guitargeek** [comment]
Ah! Absolutely, you are right that normalization with the `RooAddPdf` is not well defined in your script!

The problem is that the normalization set should be in general passed to `getVal()`. However, not doing that works most of the time because the pdfs usually know what the observables are and cache the normalization set themselves. The `RooAddPdf` doesn't do that, so calling `getVal()` gives you wrong (unnormalized) results.

In ROOT 6.24, we have implemented a warning in this case, so your script gives this output (with `option_normalize_hist=0`):
```
----------------------------
prepare model for index_process=0, signal
----------------------------
prepare model for index_process=1, Continuum
nb_events=2201.5801859878
RooRealVar::n_signal = 1.58019  L(0 - 10000) 
RooRealVar::n_Continuum = 2200  L(0 - 10000) 
nb_entries=55
-----
data : m_yy=105.5, value data=40
[#0] WARNING:Eval -- Evaluating RooAddPdf without a defined normalization set. This can lead to ambiguos coefficients definition and incorrect results. Use RooAddPdf::fixCoefNormalization(nset) to provide a normalization set for defining uniquely RooAddPdf coefficients!
value_pdf total=39.9713
n_times_pdf_signal->getVal()*bin_width=[#0] WARNING:Eval -- Evaluating RooAddPdf without a defined normalization set. This can lead to ambiguos coefficients definition and incorrect results. Use RooAddPdf::fixCoefNormalization(nset) to provide a normalization set for defining uniquely RooAddPdf coefficients!
0
n_times_pdf_continuum->getVal()*bin_width=[#0] WARNING:Eval -- Evaluating RooAddPdf without a defined normalization set. This can lead to ambiguos coefficients definition and incorrect results. Use RooAddPdf::fixCoefNormalization(nset) to provide a normalization set for defining uniquely RooAddPdf coefficients!
40
pdf signal only : 0
pdf bkg only : 40
```

If you do what the the warning suggests with `fixCoefNormalization`, or alternatively use `getVal(argset_m_yy)`, the script will give the right output:

```
----------------------------
prepare model for index_process=0, signal
----------------------------
prepare model for index_process=1, Continuum
nb_events=2201.5801859878
RooRealVar::n_signal = 1.58019  L(0 - 10000) 
RooRealVar::n_Continuum = 2200  L(0 - 10000) 
nb_entries=55
-----
data : m_yy=105.5, value data=40
value_pdf total=0.0181688
n_times_pdf_signal->getVal(argset_m_yy)*bin_width=0
n_times_pdf_continuum->getVal(argset_m_yy)*bin_width=0.0181818
pdf signal only : 0
pdf bkg only : 0.0181818
```

**Comment by marcescalier** [comment]
Jonas : thank you very much for your generous help. You are my hero for root/roofit. Everything is clear.

Unfortunately, from this discussion, it seems that I found a mistake in the way roofit is computing the nll as provided by class RooFitResult and by RooAbsPdf->createNLL.
(since I am able to reproduce the value by hand if I do a *mistake on purpose* in the formula.)

->I will prepare the minimum possible program to try to prove it (maybe there is a cartesian explanation) and will create a new github entry.
Thank you

**Comment by guitargeek** [comment]
Hi, you're welcome! Okay, then I'm looking forward to your new issue!
